kb/testing/testing-constant-flow.md

```diff
@@ -0,0 +1,101 @@
+
+References:
+- https://clang.llvm.org/docs/MemorySanitizer.html
+- https://github.com/google/sanitizers/wiki/MemorySanitizer
+- https://github.com/llvm-mirror/compiler-rt/blob/master/include/sanitizer/msan_interface.h
+
+Pros:
+- clang/memsan is already used in our CI
+- much faster than valgrind/memcheck (only 2-3x slowdown, so about 10x faster than using valgrind)
+
+Cons:
+- only works with clang on Linux, FreeBSD, NetBSD
+- being compile-time instrumentation, limited with respect to inline assembly
+- for the same reason, misses branches in toolchain-provided functions (for example long multiplication, see Kaufmann et al. [^1]) which valgrind would see
+- has an annoying bug where it [misses secret-dependent memory accesses via libc functions](https://github.com/google/sanitizers/issues/1296)
+
+Note: since 2.24.0 we're using this: see `MBEDTLS_TEST_CONSTANT_FLOW_MEMSAN` in `config.h`; components `test_memsan_constant_flow` and ``test_memsan_constant_flow_psa` in `all.sh`; and `tests/suites/test_suite_constant_time`.
+
+[^1]: When Constant-time Source Yields Variable-time Binary: Exploiting Curve25519-donna Built with MSVC 2015. Thierry Kaufmann, Hervé Pelletier, Serge Vaudenay, and Karine Villegas. 21 December 2016, Springer Science and Business Media LLC. DOI: 10.1007/978-3-319-48965-0_36. https://core.ac.uk/reader/148026376
+
+### Custom tools
+
+From looking at papers, it seems that various researchers are using various custom tools for finding leaks in cryptographic software. However it is unclear if those are publicly available, and being developed by academics mainly for their own use, it's also unclear how easy they would be to deploy on our CI infrastructure.
+
+## Static analyzers and formal guarantees
+
+In recent years there seems to be more interest from the programming languages community in eliminating side-channel attacks by providing some form of formal guarantee of safety. These methods usually involve making a new (or modifying a) language/compiler that can detect the incorrect usage of secrets.
+
+References:
+- <http://cuda.dcc.ufmg.br/flowtracker/>
+- <https://ranjitjhala.github.io/static/fact_dsl.pdf>
+- <https://eprint.iacr.org/2019/1393.pdf#section.4> (section IV, useful summary of recent ideas)
+
+Pros:
+- Often provides some guarantee of safety
+- Can sometimes transform more readable unsafe code into safe code.
+
+Cons:
+- May involve rewriting sections of constant-time code
+- Still potential for safe code but unsafe binaries
+
+## Empirical methods
+
+There appears to be some empirical testing tools available for testing constant time code. Tools such as [dudect](https://github.com/oreparaz/dudect#dudect-dude-is-my-code-constant-time) work by running the program with different inputs and trying to detect statistically different execution times. While these tools may not provide formal guarantees of safety against side-channel attacks it could be a useful, and easy to implement, metric for constant time code.
```

Comment:
My comment from yesterday `s/appears/appear/` still stands "There appear to be some empirical testing tools available" reads better

Reply:
Fixed, but for future reference, this level of typo isn't worth asking for more iterations